### PR TITLE
Add DRnR plots to PixelPhase1 DQM

### PR DESCRIPTION
**Pull request by arossi83**
#### PR description:

The main target of this PR is the addition of DRnR plots to the PixelPhase1 DQM. Everything has been packed inside a dedicated harvester and also other residuals plots creation has been moved here for consistency

#### PR validation:

Running wf 136.892 and standalone pixel online DQM client

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

not a backport and no backport forseen

### CODE REVIEW

**Comment by arossi83** [comment]
@mmusich @sroychow @dmeuser FYI

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+            DRnR_["NormRes_mean_x_Outer_PXLayer_" + layer]->Fill(me2_x->getBinContent(j, i));
+            DRnR_["NormRes_mean_y_Outer_PXLayer_" + layer]->Fill(me2_y->getBinContent(j, i));
+            DRnR_["DRnR_x_Outer_PXLayer_" + layer]->Fill(me2_x->getBinError(j, i));
+            DRnR_["DRnR_y_Outer_PXLayer_" + layer]->Fill(me2_y->getBinError(j, i));
+          }
+        }
+      }
+      for (int j : {1, 2, 3, 4, 6, 7, 8, 9}) {
+	if (me_x->getBinEntries(j, i) < minHits_){
+	  std::cout<<"Barrel Bin "<<j<<","<<i<<" - Entries "<<me2_x->getBinEntries(j, i)<<std::endl;
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc:355]: no `cout` please

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+	    DRnR_["DRnR_x_pos"]->Fill(me2_x->getBinError(j, i));
+	    DRnR_["DRnR_y_pos"]->Fill(me2_y->getBinError(j, i));
+	  }
+	  me2_x->setBinContent(j,i,me2_x->getBinError(j, i));
+	  me2_y->setBinContent(j,i,me2_y->getBinError(j, i));	
+	  me2_x->setBinEntries(me2_x->getBin(j, i),1);
+	  me2_y->setBinEntries(me2_y->getBin(j, i),1);
+	}
+	else{
+	  std::cout<<"Endcap Bin "<<j<<","<<i<<" - Entries "<<me2_x->getBinEntries(j, i)<<std::endl;
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc:442]: no `cout` please

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc [hunk 3]
```diff
@@ -81,6 +83,14 @@ namespace {
         histo[RESIDUAL_X].fill(it.resXprime, id, &iEvent);
         histo[RESIDUAL_Y].fill(it.resYprime, id, &iEvent);
 
+	if (it.resXprimeErr!=0 && it.resXprimeErr!=0){
+	  std::cout<<it.resXprime<<"  -  "<<it.resXprimeErr<<"  -  "<<it.resXprime/it.resXprimeErr<<std::endl;
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:87]: no `cout` please

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc (hunk 2 above)
Comment (arossi83) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc]: removed

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc (hunk 3 above)
Comment (arossi83) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc]: removed

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc (hunk 1 above)
Comment (arossi83) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1ResidualsExtra.cc]: removed

**Comment by mmusich** [comment]
code-checks

File: DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "DQMServices/Core/interface/DQMStore.h"
+#include "DQMServices/Core/interface/DQMEDHarvester.h"
+
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "FWCore/Framework/interface/LuminosityBlock.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h:36]: that's not needed.

File: DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+class SiPixelPhase1ResidualsExtra : public DQMEDHarvester {
+public:
+  explicit SiPixelPhase1ResidualsExtra(const edm::ParameterSet& conf);
+  ~SiPixelPhase1ResidualsExtra() override;
+
+protected:
+  // BeginJob
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h:46]: remove commented code

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc [hunk 6]
```diff
@@ -81,6 +92,13 @@ namespace {
         histo[RESIDUAL_X].fill(it.resXprime, id, &iEvent);
         histo[RESIDUAL_Y].fill(it.resYprime, id, &iEvent);
 
+        if (it.resXprimeErr != 0 && it.resXprimeErr != 0) {
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:95]: I can't understand the logic here, isn't this duplicated twice?

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc [hunk 7]
```diff
@@ -81,6 +92,13 @@ namespace {
...
+          histo[NormRes_X].fill(it.resXprime / it.resXprimeErr, id, &iEvent);
+          histo[NormRes_Y].fill(it.resYprime / it.resYprimeErr, id, &iEvent);
+          histo[DRNR_X].fill(it.resXprime / it.resXprimeErr, id, &iEvent);
```
Comment (mmusich) [inline DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:98]: how is `DRNR_X` different from `NormRes_X` ?

Comment (arossi83) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:98]: No difference at this level, the differences are introduced in the harvesting step. I separate the plots here in order to have a proper histogram name on the DQMIO root file for DRnR. Summarizing: here DRNR_X is used to fill a 2D Profile which will show NormRes mean value for each bin, in the harvesting the content of this 2D profile is changed replacing the Mean value with RMS for each bin(module).

File: DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h (hunk 5 above)
Comment (arossi83) [reply DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h]: done

File: DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h (hunk 4 above)
Comment (arossi83) [reply DQM/SiPixelPhase1Track/interface/SiPixelPhase1ResidualsExtra.h]: done

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc (hunk 7 above)
Comment (mmusich) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:98]: Can't the histogram be booked directly in the harvesting step?

**Comment by jfernan2** [comment]
please test

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc (hunk 7 above)
Comment (arossi83) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:98]: In principle we can have the MEs booked and filled in the step3 but to do that we need to implement a method top evaluate RMS of single module in a running way and be sure to be thread and Concurrent LS safe. In the way things are implemented the 2DProfile ME take care of everything and the "cost" is to have an apparently duplicated entry on step3.

**Comment by jfernan2** [comment]
+1

File: DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc (hunk 6 above)
Comment (mmusich) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc:95]: @arossi83 can you please clarify this, or remove the duplication?

Comment (arossi83) [reply DQM/SiPixelPhase1Track/plugins/SiPixelPhase1TrackResiduals.cc]: Sorry, I just realize where the duplication is, obviously is a typo. I just fixed it!

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by jfernan2** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**